### PR TITLE
feat: per-side border widths

### PR DESCRIPTION
**Pull request by Tropix126**
Adds support for individual border-widths per side. Made based on observation of figma's stroke rendering system.

## Basic Rundown

Rather than using Skia's stroking API, we instead draw the border as a second path (the "border path") with either a larger or smaller corner radius (the offset for roundness is calculated by `RectElement::border_corner_radius_offset`). We then mask out a smaller rect (the "border clip path") from the inside of the border rect.

What this gives us is the ability to offset the border rect allowing for different widths on each side.

## Current Limitations
- ~~The parser requires the user to input exactly four width values and no less currently. This means all examples and devtools are currently broken. I need to find a way to parse the widths in a clean way.~~
- ~~Devtools are broken as mentioned before.~~
- ~~Zero-width borders bleed through from under the rect if they happen to render on a subpixel.~~ This is a [skia bug](https://neugierig.org/software/chromium/notes/2010/07/clipping.html) and will be ~~fixed~~ avoided once pixel snapping is implemented. (#837)

## Some cursed showcases

Superellipse roundness with variable widths on each side. (this rendering is actually more correct than figma's, which doesn't do squircle calculations on borders :D)
image

Variable widths with variable corner radius.
image

### CODE REVIEW

**Comment by Tropix126** [comment]
Linux build is failing for reasons I can't determine. The build also took 49 minutes :p

**Comment by marc2332** [comment]
> Linux build is failing for reasons I can't determine. The build also took 49 minutes :p

Check the skia bindings

**Review by marc2332 — CHANGES_REQUESTED** [review]
Would be nice to have the border.rs example update showing this new thing right?

**Review by marc2332 — CHANGES_REQUESTED** [review]
this branch:
image

main:
image

🤔

**Comment by Tropix126** [comment]
Rendering quality has been fixed. Example has been updated as part of #889 (which is dependent on this PR).

File: crates/state/src/values/border.rs [hunk 1]
```diff
@@ -16,18 +17,60 @@ pub enum BorderStyle {
...
 impl Border {
     #[inline]
     pub fn is_visible(&self) -> bool {
-        self.width > 0. && self.fill != Fill::Color(Color::TRANSPARENT)
+        !(self.width.top == 0.0
+            && self.width.left == 0.0
+            && self.width.bottom == 0.0
+            && self.width.right == 0.0)
+            && self.fill != Fill::Color(Color::TRANSPARENT)
+            && self.style != BorderStyle::None
```
Comment (marc2332) [inline crates/state/src/values/border.rs:46]:
```suggestion
        (self.width.top != 0.0
            || self.width.left != 0.0
            || self.width.bottom != 0.0
            || self.width.right != 0.0)
            && self.fill != Fill::Color(Color::TRANSPARENT)
            && self.style != BorderStyle::None
```

**Review by marc2332 — CHANGES_REQUESTED** [review]

File: examples/border.rs
Comment (marc2332) [inline examples/border.rs:1]: Could you update this example to contain a multi side widths border?

Comment (marc2332) [reply examples/border.rs:1]: rest lgtm, will merge after this

**Review by marc2332 — CHANGES_REQUESTED** [review]

**Review by marc2332 — APPROVED** [review]

**Pull request merged**